### ISSUE TITLE
Support for uploading files in HTTP client

### ISSUE THREAD

**Issue by tindzk** [twitter/finagle#501]
I am abandoning scalaj-http in favour of Finagle. The only feature that I am missing is support for file uploads. From what I have found, there is no direct equivalent to this scalaj-http code:

```scala
val response = scalaj.http.Http(url)
  .postMulti(scalaj.http.MultiPart(
    "file",
    file.getName,
    Files.probeContentType(file.toPath),
    new FileInputStream(file),
    file.length(),
    onWrite))
```

**Comment by vkostyukov** [comment]
I think you can use [`RequestBuilder.buildFormPost(mutipart = true)`](https://github.com/twitter/finagle/blob/develop/finagle-http/src/main/scala/com/twitter/finagle/http/RequestBuilder.scala#L347) for that. See [FileElement](https://github.com/twitter/finagle/blob/develop/finagle-http/src/main/scala/com/twitter/finagle/http/RequestBuilder.scala#L30) for adding file uploads.

**Comment by tindzk** [comment]
Yes, I already saw `FileElement`. The problem is that `content` is a `Buf` and therefore the file will need to be fully loaded into memory. What I am looking for is more of a streaming solution like in scalaj-http.

**Comment by vkostyukov** [comment]
Multipart request streaming is also tricky in Finagle (because of both streaming and multipart are tricky in Finagle). Although, if you're feeling like build multipart request from scratch (not using `RequestBuilder`) it's possible to do something like:

```scala
val reader: Reader = Reader.fromFile(new File("upload.txt"))
val req: Request = Request(Version.Http11, Method.Post, "/", reader)
```

But you have to take care about the "multipart" part of this request (put content types, boundaries, etc).
